test/system/src/test/java/io/pravega/test/system/AbstractReadWriteTest.java

```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.test.system;
+
+import io.pravega.client.ClientFactory;
+import io.pravega.client.stream.EventRead;
+import io.pravega.client.stream.EventStreamReader;
+import io.pravega.client.stream.EventStreamWriter;
+import io.pravega.client.stream.EventWriterConfig;
+import io.pravega.client.stream.ReaderConfig;
+import io.pravega.client.stream.ReinitializationRequiredException;
+import io.pravega.client.stream.impl.JavaSerializer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import lombok.Cleanup;
+import lombok.extern.slf4j.Slf4j;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Abstract class that provides convenient event read/write methods for testing purposes.
+ */
+@Slf4j
+abstract class AbstractReadWriteTest {
+
+    private static final int READ_TIMEOUT = 1000;
+
+    protected void writeDummyEvents(ClientFactory clientFactory, String streamName, int totalEvents) {
+        @Cleanup
+        EventStreamWriter<String> writer = clientFactory.createEventWriter(streamName, new JavaSerializer<>(),
+                EventWriterConfig.builder().build());
+        for (int i = 0; i < totalEvents; i++) {
+            writer.writeEvent(streamName + String.valueOf(i)).join();
+            log.debug("Writing event: {} to stream {}.", streamName + String.valueOf(i), streamName);
+        }
+    }
+
+    protected List<CompletableFuture<Integer>> readEventFutures(ClientFactory client, String rGroup, int numReaders) {
+        List<EventStreamReader<String>> readers = new ArrayList<>();
+        for (int i = 0; i < numReaders; i++) {
+            readers.add(client.createReader(String.valueOf(i), rGroup, new JavaSerializer<>(), ReaderConfig.builder().build()));
+        }
+
+        return readers.stream().map(r -> CompletableFuture.supplyAsync(() -> readEvents(r))).collect(toList());
+    }
+
+    protected <T> int readEvents(EventStreamReader<T> reader) {
+        EventRead<T> event;
+        int validEvents = 0;
+        try {
+            do {
+                event = reader.readNextEvent(READ_TIMEOUT);
+                log.debug("Read event result in readEvents: {}.", event.getEvent());
+                if (event.getEvent() != null) {
+                    validEvents++;
+                }
+            } while (event.getEvent() != null || event.isCheckpoint());
+        } catch (ReinitializationRequiredException e) {
+            throw new RuntimeException(e);
```

Comment:
Explain to me what the expected behavior is. The one test case this is currently targeting deals with stream cuts and a reader will be forced reinitialization in the case the group is restored to a stream cut. Do you really want to propagate the exception rather than reinitialize the reader? Let's think about the concrete expected behavior of this method.

Reply:
In the current implementation of the test, `ReinitializationRequiredException` is not expected as readers within a group are not disrupted by a `resetReaderGroup` while they are reading. In fact, inspecting the test logs I have not found any `ReinitializationRequiredException`. For this reason `ReinitializationRequiredException` was treated like this. Anyway, in the future we may want to test the behavior of readers when handling a `ReinitializationRequiredException`. Thus, in the current version of `readEvents` a reader tolerates this exception and attempts to continue reading after resetting takes place.